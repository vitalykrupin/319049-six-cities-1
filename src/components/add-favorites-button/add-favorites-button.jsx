import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {withRouter} from 'react-router-dom';
import {Operation} from '../../reducer/user/user';
import {getAuthorizationStatus} from '../../reducer/user/selectors';

class AddFavoritesButton extends PureComponent {
  constructor(props) {
    super(props);

    this._handleButtonClick = this._handleButtonClick.bind(this);
  }

  render() {
    const {className, fromRoom, place} = this.props;
    return <button className={`${className}__bookmark-button button ${this._getButtonClass(place.isFavorite)}`} type="button" onClick={() => this._handleButtonClick()}>
      <svg className={this._getIconClass(fromRoom, place.isFavorite)} width={fromRoom ? `31` : `18`} height={fromRoom ? `33` : `19`}>
        <use xlinkHref="#icon-bookmark"/>
      </svg>
      <span className="visually-hidden">To bookmarks</span>
    </button>;
  }

  _getButtonClass(isFavorite) {
    return isFavorite ? `place-card__bookmark-button--active` : ``;
  }

  _getIconClass(fromRoom, isFavorite) {
    if (isFavorite && fromRoom) {
      return `property__bookmark-icon--active`;
    } if (!isFavorite && fromRoom) {
      return `property__bookmark-icon`;
    } else {
      return `place-card__bookmark-icon`;
    }
  }

  _handleButtonClick() {
    const {addToFavorites, deleteFromFavorites, place, isAuthorizationRequired} = this.props;

    if (!isAuthorizationRequired) {
      if (place.isFavorite) {
        deleteFromFavorites(place.id);
      } else {
        addToFavorites(place.id);
      }
    } else {
      const {history} = this.props;
      history.push(`/login`);
    }
  }
}

AddFavoritesButton.propTypes = {
  place: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    isPremium: PropTypes.bool.isRequired,
    price: PropTypes.number.isRequired,
    rating: PropTypes.number.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    type: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    images: PropTypes.array.isRequired,
    goods: PropTypes.array.isRequired,
    bedrooms: PropTypes.number.isRequired,
    maxAdults: PropTypes.number.isRequired,
    host: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired,
    city: PropTypes.shape({
      name: PropTypes.string.isRequired,
      location: PropTypes.object.isRequired,
    }).isRequired,
  }).isRequired,
  fromRoom: PropTypes.bool,
  addToFavorites: PropTypes.func.isRequired,
  deleteFromFavorites: PropTypes.func.isRequired,
  isAuthorizationRequired: PropTypes.bool.isRequired,
  history: PropTypes.object,
  className: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  isAuthorizationRequired: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  addToFavorites: (id) => dispatch(Operation.addToFavorites(id)),
  deleteFromFavorites: (id) => dispatch(Operation.deleteFromFavorites(id))
});

export {AddFavoritesButton};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(AddFavoritesButton));
