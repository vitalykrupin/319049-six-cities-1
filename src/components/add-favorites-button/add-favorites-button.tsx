import * as React from 'react';
import {connect} from 'react-redux';
import {withRouter} from 'react-router-dom';
import {Operation} from '../../reducer/user/user';
import {getAuthorizationStatus} from '../../reducer/user/selectors';
import {Place} from '../../types';

interface Props {
  place: Place,
  fromRoom: boolean,
  history: {push: (path: string) => void},
  className: string,
  isAuthorizationRequired: boolean,
  addToFavorites: (id: number) => void,
  deleteFromFavorites: (id: number) => void
}

class AddFavoritesButton extends React.PureComponent<Props> {
  constructor(props) {
    super(props);

    this._handleButtonClick = this._handleButtonClick.bind(this);
  }

  render() {
    const {
      className,
      fromRoom,
      place
    } = this.props;

    return <button className={`${className}__bookmark-button button ${this._getButtonClass(place.isFavorite)}`} type="button" onClick={() => this._handleButtonClick()}>
      <svg className={this._getIconClass(fromRoom, place.isFavorite)} width={fromRoom ? `31` : `18`} height={fromRoom ? `33` : `19`}>
        <use xlinkHref="#icon-bookmark"/>
      </svg>
      <span className="visually-hidden">To bookmarks</span>
    </button>;
  }

  _getButtonClass(isFavorite) {
    return isFavorite ? `place-card__bookmark-button--active` : ``;
  }

  _getIconClass(fromRoom, isFavorite) {
    if (isFavorite && fromRoom) {
      return `property__bookmark-icon--active`;
    } if (!isFavorite && fromRoom) {
      return `property__bookmark-icon`;
    } else {
      return `place-card__bookmark-icon`;
    }
  }

  _handleButtonClick() {
    const {
      addToFavorites,
      deleteFromFavorites,
      place,
      isAuthorizationRequired
    } = this.props;

    if (!isAuthorizationRequired) {
      if (place.isFavorite) {
        deleteFromFavorites(place.id);
      } else {
        addToFavorites(place.id);
      }
    } else {
      const {history} = this.props;
      history.push(`/login`);
    }
  }
}

const mapStateToProps = (state) => ({
  isAuthorizationRequired: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  addToFavorites: (id) => dispatch(Operation.addToFavorites(id)),
  deleteFromFavorites: (id) => dispatch(Operation.deleteFromFavorites(id))
});

export {AddFavoritesButton};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(AddFavoritesButton));
