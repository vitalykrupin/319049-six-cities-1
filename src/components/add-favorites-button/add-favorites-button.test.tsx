import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {AddFavoritesButton} from './add-favorites-button';

const place = {
  id: 1,
  title: `Strange place`,
  isPremium: true,
  price: 1200,
  rating: 1.9,
  isFavorite: false,
  description: ``,
  type: `Apartment`,
  previewImage: ``,
  images: [``],
  goods: [``],
  bedrooms: 2,
  maxAdults: 4,
  host: {
    id: 3,
    name: `Alice`,
    isPro: false,
    avatarUrl: `path`
  },
  location: {
    latitude: 12,
    longitude: 87,
    zoom: 11,
  },
  city: {
    name: `Berlin`,
    location: {
      latitude: 51,
      longitude: 7,
      zoom: 11,
    },
  },
};

describe(`AddFavoritesButton`, () => {
  it(`renders correctly`, () => {
    const tree = renderer
      .create(
          <AddFavoritesButton
            place={place}
            addToFavorites={jest.fn()}
            deleteFromFavorites={jest.fn()}
            isAuthorizationRequired={true}
            className={`property`}
            fromRoom={false}
            history={{push: jest.fn()}}
          />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
