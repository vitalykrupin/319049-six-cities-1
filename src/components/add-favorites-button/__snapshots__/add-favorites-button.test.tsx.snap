// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFavoritesButton renders correctly 1`] = `
<button
  className="property__bookmark-button button "
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height="19"
    width="18"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;
